### DISCUSSION TITLE
React Native 0.73 版本支持

### DISCUSSION THREAD

**Discussion by zhiqingchen** [NervJS/taro#14128] [RFC]
- 提案时间: (2023-12-12)
- 影响版本: (4.0)

## 概述

将默认的 RN 版本切换至 0.73，expo 版本切换至 [50](https://blog.expo.dev/expo-sdk-50-beta-is-now-available-8e6c74e4e13d)。
参考[React Native 0.71 版本支持](https://github.com/NervJS/taro/discussions/13148)

## 动机

[React Native 0.73 版本已于 2023-12-6 正式发布](https://reactnative.dev/blog/2023/12/06/0.73-debugging-improvements-stable-symlinks)。
Taro 将与 RN 社区保持同步，将默认的 RN 版本设置为 0.73。

## 使用案例

将 Taro 升级至最新版，将原生代码根据相关pr进行变更。

### 依赖变更

todo

### Taro 主仓库升级至 0.73

升级内容：

0. 将不兼容的RN依赖库进行升级，并验证其功能可用性。
6. 将 `react-native` 进行版本更新。
7. 将 metro、@react-native-community/cli-* 相关依赖与 0.73 使用的版本进行同步。
8. 将默认安装的RN版本进行更新 https://github.com/NervJS/taro/blob/next/packages/taro-cli/src/presets/platforms/rn.ts#L33
9. @rnx-kit/metro-resolver-symlinks 删除（0.73官方会支持）
10. react-native-syan-image-picker 改为 expo-image-picker

### Taro Native Shell 升级至 0.73

分支：https://github.com/NervJS/taro-native-shell/tree/0.73.0

参考 [upgrade-helper](https://react-native-community.github.io/upgrade-helper/?from=0.72.0&to=0.73.0)

### Taro Project Template 升级至 0.73

0. 更新代码同 Taro Native Shell 升级至 0.73。

### Taro Playground 升级至 0.73

分支：

https://github.com/wuba/taro-playground/tree/releases/v1.11.0

升级内容：

0. 版本号改为 v1.10.0。
0. 更新代码同 Taro Native Shell 升级至 0.73。

## 缺陷

- 0.73 版本发布时间较短，及时跟进 0.73 问题。

## 替代选择

N/A

## 适配策略

- Taro 主工程默认使用 0.73，保证旧版本 RN 仍正常使用新版本的 Taro。
- 有些包（如 metro）可能无法向下兼容，则需要用户使用 resolution 进行版本锁定，需要在版本更新时说明。

### metro.config.js

```
const {getDefaultConfig, mergeConfig} = require('@react-native/metro-config');
const { getMetroConfig } = require('@tarojs/rn-supporter')

/**
 * Metro configuration
 * https://facebook.github.io/metro/docs/configuration
 *
 * @type {import('metro-config').MetroConfig}
 */
const config = {};

module.exports = (async function (){
    return mergeConfig(getDefaultConfig(__dirname), await getMetroConfig(), config)
})()
```

### package.json

```
    "start": "react-native start --custom-log-reporter-path @tarojs/rn-supporter/TerminalReporter.js",
```

**Comment by zhiqingchen** [comment]
## Taro Native Shell 升级至 0.73 测试

对应分支

https://github.com/NervJS/taro-native-shell/tree/0.73.0

下载现有包进行验证

https://github.com/NervJS/taro-native-shell/releases/tag/v0.73.0-0

修改代码后

```
# 10 为自增数字
git tag v0.73.0-3
git push origin v0.73.0-3
```

查看打包是否正常进行

https://github.com/NervJS/taro-native-shell/actions

**Reply by zhiqingchen** [reply]
遗留问题：

- [ ] https://github.com/expo/expo/issues/25979

**Comment by zhiqingchen** [comment]
## Taro 主仓库升级至 0.73

对应分支

https://github.com/NervJS/taro/tree/feat/react-native-73

包安装

4.0.0-canary.1

### 测试要点

无

**Reply by zhiqingchen** [reply]
相关问题：

https://github.com/react-native-community/cli/pull/2024

**Comment by zhiqingchen** [comment]
## Taro Project Template 升级至 0.73 测试

对应分支

https://github.com/zhiqingchen/taro-project-templates/tree/v4.0

```
# 全局安装beta版本cli
yarn global add @tarojs/cli@canary

# 通过模板进行初始
taro init --template-source github:zhiqingchen/taro-project-templates#v4.0

# 初始化后安装依赖
yarn upgradePeerdeps
```

1. 修改 assemble_ios_debug.yml 配置
   1. APP_ID: ${{secrets.APP_ID}}
   2. TEAM_ID: ${{secrets.TEAM_ID}}
   3. PROVISIONING_PROFILE_SPECIFIER: ${{secrets.DEBUG_PROVISIONING_PROFILE_SPECIFIER}}
   4. CODE_SIGN_IDENTITY: Apple Development
2. 修改 assemble_ios_release.yml 配置
   1. APP_ID: ${{secrets.APP_ID}}
   2. TEAM_ID: ${{secrets.TEAM_ID}}
   3. PROVISIONING_PROFILE_SPECIFIER: ${{secrets.RELEASE_PROVISIONING_PROFILE_SPECIFIER}}
   4. CODE_SIGN_IDENTITY: Apple Distribution

```
# 将源设置为 shell 工程
git remote add origin git@github.com:NervJS/taro-native-shell.git

# 10 为自增数字
git tag v0.73.0-beta.0
git push origin v0.73.0-beta.0
```

查看打包是否正常进行

https://github.com/NervJS/taro-native-shell/actions

参考分支

https://github.com/NervJS/taro-native-shell/tree/test/template-0.73-0

**Comment by zhiqingchen** [comment]
## Taro Playground 升级至 0.73 测试

对应分支

https://github.com/wuba/taro-playground/tree/releases/v1.11.0

包下载
https://github.com/wuba/taro-playground/releases/tag/v1.11.0-beta.1

**Comment by zhiqingchen** [comment]
## pnpm 支持

### template 缺失依赖

- @tarojs/rn-supporter
- @tarojs/runtime-rn
- @tarojs/components-rn

* @react-native-community/cli-platform-android@^12.3.0
* ~~@react-native-community/cli-platform-ios@^12.3.0~~
* @react-native/gradle-plugin@^0.73.0

### 多余依赖

- "@babel/preset-env": "^7.20.0",

### upgradePeerdeps 脚本更新

- Y 改成 -P

### github action 更新

### 文档更新

- readme
- docs

### 壳工程更新

**Comment by zhpd** [comment]
升级到React Native 0.74 后，会报错，脚本无法编译，大概判断问题是在const { getMetroConfig } = require('@tarojs/rn-supporter')，这个方法里面，能帮忙定位处理下吗

**Reply by zhiqingchen** [reply]
是通过什么版本升级上来的？

**Reply by zhpd** [reply]
> 是通过什么版本升级上来的？

Taro 4.0.0-beta.92升级的，原来的0.73正常可以用，升级到0.74后脚本无法正常编译，怀疑是这个影响的const { getMetroConfig } = require('@tarojs/rn-supporter')，去掉这个，rn就可以正常运行

**Comment by wuweikd** [comment]
请问计划什么时候升级呢

**Comment by zhiqingchen** [comment]
已合并
